core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java

```diff
@@ -156,7 +156,11 @@ public static void run(Application application, Class<? extends QuarkusApplicati
                 }
             }
         } catch (Exception e) {
-            Throwable rootCause = ExceptionUtil.getRootCause(e);
+            // Here if ConfigurationException is one step up on the chain we want to return it
+            // because it wraps ConfigValidationException from io.smallrye.config
+            Throwable rootCause = e.getCause() instanceof ConfigurationException
+                    ? e.getCause()
+                    : ExceptionUtil.getRootCause(e);
```

Comment:
I'm a bit concerned this will start failing again as soon as there's a change in how we wrap exceptions... E.g. the `ConfigurationException` could be `e`, or ` e.getCause()`, or `e.getCause().getCause()`, or...

I think it would be safer to  just make `ConfigurationException` a special case and look for it through the whole stack....

I'd suggest to keep this the way it was:

```suggestion
            Throwable rootCause = ExceptionUtil.getRootCause(e);
```

... And below, do something like this:

```diff
-                 } else if (rootCause instanceof ConfigurationException) {
+                 } else if (ExceptionUtil.isAnyCauseInstanceOf(e, ConfigurationException.class)) {
```

(`ExceptionUtil.isAnyCauseInstanceOf` would have to be implemented, of course)

Reply:
I added the changes that you suggested, wrote test for new method and tested this on some project - everything seems fine. The only thing that's different now is the exception message printed at application startup failure, now it prints rootCause message.

> I think I should probably open a follow-up issue to implement some kind of listener (similar to ApplicationStateNotification -- maybe we could use that?) to register the information "the application managed to create and validate configuration, error reporting code can use it safely". And we'd get rid of this attempt at guessing this information from very incomplete data.

If you believe such thing might be useful then it sounds good to me :) Maybe I will have a chance to look into this followup issue and try myself

Reply:
> If you believe such thing might be useful then it sounds good to me :)

Done: https://github.com/quarkusio/quarkus/issues/37846

> Maybe I will have a chance to look into this followup issue and try myself

That would be nice :) I have no idea how complex that will be, though :D